homeassistant/helpers/entity_component.py

```diff
@@ -391,47 +357,117 @@ def async_add_entities(self, new_entities, update_before_add=False):
         if not new_entities:
             return
 
-        @asyncio.coroutine
-        def async_process_entity(new_entity):
-            """Add entities to StateMachine."""
-            new_entity.parallel_updates = self.parallel_updates
-            ret = yield from self.component.async_add_entity(
-                new_entity, self, update_before_add=update_before_add
-            )
-            if ret:
-                self.platform_entities.append(new_entity)
+        component_entities = set(entity.entity_id for entity
+                                 in self.component.entities)
 
-        tasks = [async_process_entity(entity) for entity in new_entities]
+        tasks = [
+            self._async_add_entity(entity, update_before_add,
+                                   component_entities)
+            for entity in new_entities]
 
         yield from asyncio.wait(tasks, loop=self.component.hass.loop)
         self.component.async_update_group()
 
         if self._async_unsub_polling is not None or \
            not any(entity.should_poll for entity
-                   in self.platform_entities):
+                   in self.entities.values()):
             return
 
         self._async_unsub_polling = async_track_time_interval(
             self.component.hass, self._update_entity_states, self.scan_interval
         )
 
+    @asyncio.coroutine
+    def _async_add_entity(self, entity, update_before_add, component_entities):
+        """Helper method to add an entity to the platform."""
+        if entity is None:
+            raise ValueError('Entity cannot be None')
+
+        # Do nothing if entity has already been added based on unique id.
+        if entity in self.component.entities:
+            return
+
+        entity.hass = self.component.hass
+        entity.platform = self
```

Comment:
I think that is not needed or we should not do it for our software designe. We should add the platform to `setup_platform(..., platform=None)` so we need not store the entities anymore in `hass.data` for own service handling.

Reply:
I don't want to change the method signature of platform setup. That will break all platforms.

One thing we _could_ do (and it feels like a hack), is to make `EntityPlatform` callable and behave like `async_add_devices` for async setup.

Reply:
Current responsibilities:

- EntityComponent: responsible for managing platforms, group of all entities in all platforms, discovery, extracting entities from services
- EntityPlatform: responsible for managing the entities of a platform and keeping them up to date
- Entity: responsible for representing an entity in the state machine

EntityPlatform already knows about it's parent EntityComponent to notify that the group needs updating.

The alternative to the entity not knowing the platform it belongs to is using an event to have the entity signal that something should happen. However, in the case of remove, it feels like we're abusing an event system: it is a call for action instead of something that did happen.

Reply:
Yes, you are right